cluster-autoscaler/core/scale_up.go

```diff
@@ -322,3 +362,90 @@ func addAutoprovisionedCandidates(context *AutoscalingContext, nodeGroups []clou
 		}
 	}
 }
+
+func calculateCPUAndMemory(nodeGroups []cloudprovider.NodeGroup, nodeInfos map[string]*schedulercache.NodeInfo) (int64, int64, []error) {
+	var coresTotal int64
+	var memoryTotal int64
+	var errs []error = make([]error, 0)
+	for _, nodeGroup := range nodeGroups {
+		currentSize, err := nodeGroup.TargetSize()
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Failed to get node group size: %v", err))
+		}
+		nodeInfo, found := nodeInfos[nodeGroup.Id()]
+		if !found {
+			errs = append(errs, fmt.Errorf("No node info for: %s", nodeGroup.Id()))
+		}
+		if currentSize > 0 {
+			nodeCPU, nodeMemory, err := getNodeCPUAndMemory(nodeInfo)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("Failed to get node resources: %v", err))
+			}
+			coresTotal = coresTotal + int64(currentSize)*nodeCPU
+			memoryTotal = memoryTotal + int64(currentSize)*nodeMemory
+		}
+	}
+
+	return coresTotal, memoryTotal, errs
+}
+
+func applyCPUAndMemoryLimit(newNodes int, coresTotal, memoryTotal, maxCoresTotal, maxMemoryTotal int64, nodeInfo *schedulercache.NodeInfo) (int, errors.AutoscalerError) {
+	newNodeCPU, newNodeMemory, err := getNodeCPUAndMemory(nodeInfo)
+	if err != nil {
+		// This is not very elegant, but it allows us to proceed even if we're
+		// unable to compute cpu/memory limits (not breaking current functionality)
+		glog.Errorf("Failed to get node resources: %v", err)
+		return newNodes, nil
+	}
+	if coresTotal+newNodeCPU*int64(newNodes) > maxCoresTotal {
+		glog.V(1).Infof("Capping size to max cluster cores (%d)", maxCoresTotal)
+		newNodes = int((maxCoresTotal - coresTotal) / newNodeCPU)
+		if newNodes < 1 {
+			// This should never happen, as we already check that
+			// at least one node will fit when considering nodegroup
+			return 0, errors.NewAutoscalerError(
+				errors.TransientError,
+				"max cores already reached")
+		}
+	}
+	if memoryTotal+newNodeMemory*int64(newNodes) > maxMemoryTotal {
+		glog.V(1).Infof("Capping size to max cluster memory allowed (%d)", maxMemoryTotal)
+		newNodes = int((maxMemoryTotal - memoryTotal) / newNodeMemory)
+		if newNodes < 1 {
+			// This should never happen, as we already check that
+			// at least one node will fit when considering nodegroup
+			return 0, errors.NewAutoscalerError(
+				errors.TransientError,
+				"max memory already reached")
+		}
+	}
+	return newNodes, nil
+}
+
+const (
+	// Megabyte is 2^20 bytes.
+	Megabyte float64 = 1024 * 1024
+)
+
+func getNodeCPUAndMemory(nodeInfo *schedulercache.NodeInfo) (int64, int64, error) {
```

Comment:
same here. it is very unlikely that there will be no cpu/mem in node

Reply:
True, and I suppose if we did get a corrupted object somehow, perhaps we should break out of the main loop. Still, I'd feel better checking for it:)